qualtran/bloqs/rotations/hamming_weight_phasing.py

```diff
@@ -74,24 +78,77 @@ class HammingWeightPhasing(GateWithRegisters):
     bitsize: int
     exponent: float = 1
     eps: SymbolicFloat = 1e-10
+    phase_oracle: QvrInterface = attrs.field()
+
+    @phase_oracle.default
+    def _default_qvr(self):
+        return QvrZPow(
+            Register('out', QFxp(self._hamming_weight_bitsize, 0)),
+            gamma=self.exponent / 2,
+            eps=self.eps,
+        )
 
     @cached_property
     def signature(self) -> 'Signature':
-        return Signature.build_from_dtypes(x=QUInt(self.bitsize))
+        return Signature([Register("x", QUInt(self.bitsize)), *self.phase_oracle.extra_registers])
+
+    @classmethod
+    def via_phase_gradient(cls, bitsize: int, exponent: float = 1, eps: SymbolicFloat = 1e-10):
+        r"""Factory method for `HammingWeightPhasing` using a phase gradient state.
+
+        In this variant of Hamming Weight Phasing, instead of directly synthesizing $O(\log_2 (\texttt{bitsize}))$
+        rotations on the Hamming weight register we synthesize the rotations via an addition into the
+        phase gradient register. See reference [1] for more details on this technique.
+
+        Note: For most reasonable values of `bitsize` and `eps`, the naive `HammingWeightPhasing` would
+        have better constant factors than `HammingWeightPhasingViaPhaseGradient`. This is because, in
+        general, the primary advantage of using phase gradient is to reduce the complexity from
+        $O(n * \log(1/ \texttt{eps} ))$ to $O(\log^2(1/ \texttt{eps} ))$ (the phase gradient register is of size
+        $O(\log(1/\texttt{eps}))$ and a scaled addition into the target takes $(b_{grad} - 2)(\log(1/\texttt{eps}) + 2)$).
+        Therefore, to apply $n$ individual rotations on a target register of size $n$, the complexity is
+        independent of $n$ and is essentially a constant (scales only with $log(1/\texttt{eps})$).
+        However, for the actual constant values to be better, the value of $n$ needs to be
+        $> \log(1/\texttt{eps})$. In the case of hamming weight phasing, $n$ corresponds to the hamming weight
+        register which itself is $\log(\texttt{bitsize})$. Thus, as `eps` becomes smaller, the required
+        value of $\texttt{bitsize}$, for the phase gradient version to become more performant, becomes
+        larger.
+
+        Args:
+            bitsize: Size of input register to apply `Z ** exponent` to.
+            exponent: The exponent of `Z ** exponent` to be applied to each qubit in the input register.
+            eps: Accuracy of synthesizing the Z rotations.
+
+        Registers:
+            x : Input THRU register of size `bitsize`, to apply `Z**exponent` to.
+            phase_grad : Phase gradient THRU register of size `O(log2(1/eps))`, to be used to
+                apply the phasing via addition.
+
+        References:
+            1. [Compilation of Fault-Tolerant Quantum Heuristics for Combinatorial Optimization]
```

Comment:
I know you copied this, but remove `1. ` from the list of references. We don't number them.